thirdparty/DiskANN/src/pq_flash_index.cpp

```diff
@@ -872,6 +874,120 @@ namespace diskann {
+    frontier_read_reqs.reserve(2 * beam_width);
+    std::vector<std::pair<unsigned, std::pair<unsigned, unsigned *>>>
+        cached_nhoods;
+    cached_nhoods.reserve(2 * beam_width);
+
+    std::unordered_map<_u64, std::vector<_u64>> nodes_in_sectors_to_visit;
+    // Switch to brute force when filter rate is high
+    if (!bitset_view.empty() && bitset_view.count() > bitset_view.size() * kDiskAnnBruteForceFilterRate) {
+      knowhere::ResultMaxHeap<float, _u64>  max_heap(k_search);
+      // TODO: maybe we can pipeline here
+      assert(bitset_view.size() == num_points);
+      for (_u64 id = 0; id < num_points; ) {
+        // gathering information about reading which nodes and which sector
+        if (!bitset_view.test(id)) {
+          const _u64 sector_offset = get_node_sector_offset(id);
+          if (nodes_in_sectors_to_visit.find(sector_offset) == nodes_in_sectors_to_visit.end()) {
+            while (id < num_points && get_node_sector_offset(id) == sector_offset) {
+              if (!bitset_view.test(id)) {
+                if (coord_cache.find(id) != coord_cache.end()) {
+                  float dist = dist_cmp(query, coord_cache.at(id), (size_t) aligned_dim);
+                  max_heap.Push(dist, id);
+                } else {
+                  nodes_in_sectors_to_visit[sector_offset].push_back(id);
+                }
+              }
+              ++id;
+            }
+          } else {
+            LOG_KNOWHERE_ERROR_ << "Should not be visited twice";
+          }
+        } else {
+          ++id;
+        }
+        if (id < num_points && nodes_in_sectors_to_visit.size() < beam_width) continue;
+
+        // perform I/O
+        for (const auto& [sector_offset, ids_in_sectors] : nodes_in_sectors_to_visit) {
+          frontier_read_reqs.emplace_back(
+              sector_offset, read_len_for_node,
+              sector_scratch + sector_scratch_idx * read_len_for_node);
+          ++sector_scratch_idx;
+          if (stats != nullptr) {
+            stats->n_4k++;
+            stats->n_ios++;
+          }
+        }
+        io_timer.reset();
+#ifdef USE_BING_INFRA
+        reader->read(frontier_read_reqs, ctx, true);  // async reader windows.
+#else
+        reader->read(frontier_read_reqs, ctx);  // synchronous IO linux
+#endif
+        if (stats != nullptr) {
+          stats->io_us += (double) io_timer.elapsed();
+        }
+
+        T *node_fp_coords_copy = data_buf;
+        for (const auto& req : frontier_read_reqs) {
+          const _u64 sector_offset = req.offset;
+          char* sector_buf = reinterpret_cast<char*>(req.buf);
+          for (const auto cur_id : nodes_in_sectors_to_visit[sector_offset]) {
+            char* node_buf = get_offset_to_node(sector_buf, cur_id);
+            memcpy(node_fp_coords_copy, node_buf, disk_bytes_per_point); // Do we really need memcpy here?
+            float dist = dist_cmp(query, node_fp_coords_copy, (size_t) aligned_dim);
```

Comment:
Why we need this copy?

Reply:
For different alignment memory, the trade-offs between memcpy and calculation speedup should be revised later